### ISSUE TITLE
"Lock not held" errors reported in console when exporting screenshot from Game History panel

### ISSUE THREAD

**Issue by ssoloff** [triplea-game/triplea#1445]
**Environment**
- TripleA 1.9.0.0.3525
- Fedora 24 (Linux 4.8.15-200.fc24.x86_64)

**Reproduction Steps**
1. Start TripleA.
1. Select map (I used _Classic: Iron Blitz 3rd Edition_, but I observed the issue with other maps, as well).
1. Start local game.
1. Open the Game History panel (**Game > Show history**).
1. Open the context menu on the Game History panel and select **Save Screenshot**.
1. Select a location for the saved screenshot and click **Save**.

**Expected Behavior**
No errors should be reported in the console.

**Actual Behavior**
Multiple identical stack traces are displayed in the console.  A partial stack trace follows:

    java.lang.Exception: Lock not held
        at games.strategy.engine.data.GameData.ensureLockHeld(GameData.java:135)
        at games.strategy.engine.data.GameData.getRelationshipTracker(GameData.java:438)
        at games.strategy.triplea.ui.screen.drawable.BattleDrawable.draw(BattleDrawable.java:40)
        at games.strategy.triplea.ui.screen.Tile.draw(Tile.java:124)
        at games.strategy.triplea.ui.screen.Tile.getImage(Tile.java:80)
        at games.strategy.triplea.ui.MapPanel.print(MapPanel.java:476)
        at games.strategy.triplea.ui.menubar.ExportMenu.saveScreenshot(ExportMenu.java:215)
        . . .

The complete [console log](https://github.com/triplea-game/triplea/files/698059/console.log.zip) (with 457 occurrences of the above stack trace) is attached.

**Notes**
- The screenshot is exported without any issues.
- Immediately repeating the attempt to export a screenshot from the Game History panel results in no errors being reported in the console.

**Comment by ssoloff** [comment]
I did some brief analysis on this issue...  It appears the reason why the console errors occur only when exporting a screenshot from the Game History panel and not from the main menu is because of the following line in `TripleAFrame#showHistory()`:

    clonedGameData.testLocksOnRead();

This method sets a flag that allows the checks in `GameData#ensureLockHeld()` to be executed (and subsequently print the error to the console).  If I add the following statement:

    gameData.testLocksOnRead();

immediately before the call to `saveScreenshot()` in `ExportMenu#addSaveScreenshot()`, I can reproduce the console errors when exporting the screenshot from the main menu.

The reason why immediately repeating the attempt to export the screenshot from the Game History panel results in no console errors is because the `Tile#draw()` method (which ultimately leads to the call to `GameData#ensureLockHeld()`) is only invoked if the tile is dirty.  In my reproduction steps, I did not do anything to trigger a tile to become dirty.  Presumably, if I changed the game state such that at least one tile was modified, I would observe a corresponding console error when repeating the attempt to export the screenshot.

I get the impression that the `GameData` read lock is supposed to be held while exporting a screenshot (since the game data is ultimately being accessed in a non-atomic manner during the export).  And that the call to `GameData#testLocksOnRead()` in the Game History panel uncovered this missing synchronization.  Are missing calls to `GameData#acquireReadLock()` and `GameData#releaseReadLock()` around the export screenshot code the correct fix for this issue?

**Comment by simon33-2** [comment]
I can't find the save screenshot button you refer to.

@ssoloff

**Comment by ssoloff** [comment]
Sorry, the above repro steps were created before PR #1450 was merged.  That PR changed the menu item label from **Save Screenshot** to **Export Map Snapshot** (see the screenshot attached to the PR).  I will update the repro steps in this issue to reflect the current UI.

**Comment by simon33-2** [comment]
Ok, I can't duplicate the problem. Perhaps another change has fixed it?

**Comment by ssoloff** [comment]
I just synced up with the latest commit on `master` (commit b07d7cf, tag 1.9.0.1.3650), and the problem still exists.

I didn't state this explicitly in the repro steps, but did you open the console manually (i.e. via **Debug > Show Console**)?  The error message are logged as text rather than as exceptions, so the console does not open automatically.  (I actually observed it first in my Eclipse console while running TripleA from within Eclipse; otherwise I probably wouldn't have ever noticed it because I wouldn't have thought to manually open the TripleA console after exporting a screenshot.)

**Comment by ssoloff** [comment]
If there's still an issue reproducing the problem, I've attached a [screencast](https://github.com/triplea-game/triplea/files/737968/screencast-1.9.0.0.3700.zip) that demonstrates the problem on version 1.9.0.0.3700.

**Comment by ssoloff** [comment]
I confirmed this issue is no longer present in build 1.9.0.0.3712 with the merge of #1477.  I've attached a [screencast](https://github.com/triplea-game/triplea/files/758153/screencast-1.9.0.0.3712.zip) that demonstrates the error messages are no longer displayed in the console (compare with the previous screencast above).

@ron-murhammer Even though I'm the ticket author, I'm not closing this issue myself because I'm not sure if you guys require someone other than the contributor to verify the fix.
